compiler/rustc_const_eval/src/interpret/terminator.rs

```diff
@@ -372,14 +372,7 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         };
         if let (Some(caller), Some(callee)) = (pointee_ty(caller.ty)?, pointee_ty(callee.ty)?) {
             // This is okay if they have the same metadata type.
-            let meta_ty = |ty: Ty<'tcx>| {
-                let (meta, only_if_sized) = ty.ptr_metadata_ty(*self.tcx, |ty| ty);
-                assert!(
-                    !only_if_sized,
-                    "there should be no more 'maybe has that metadata' types during interpretation"
-                );
-                meta
-            };
+            let meta_ty = |ty: Ty<'tcx>| ty.ptr_metadata_ty(*self.tcx, |ty| ty);
```

Comment:
(existing issue) why do we not normalize here? feels wrong. I expect that this can result in an ICE when using a type with an alias as a tail.

Reply:
Parts of the interpreter are called on generic code. So it'd be better to handle this with `TooGeneric`, like we already do elsewhere.

I just don't understand why we need to normalize here at all, since all types go through `subst_from_current_frame_and_normalize_erasing_regions` so they should already be normalized.

Reply:
In particular, we have a bunch of matches on `ty.kind()` above this line that seem like they may miss the types they check for if this can be non-normalized?

Reply:
I think the issue might be here:

https://github.com/rust-lang/rust/blob/16ba56c24238b2f44e0a97da760c83d8795abd80/compiler/rustc_const_eval/src/interpret/terminator.rs#L686-L692

We project to a field but then do not normalize again.

Reply:
> All `TyAndLayout` floating around in the interpreter must already be normalized. So adding normalization here is _not_ a good idea, it just masks an issue somewhere else where we are failing to normalize.

Even if all types are normalized, a field of a normalized ADT can still have an unnormalized type.

For example here

```rust
trait Project {
    type Assoc;
}

struct Foo<T: Project> {
    x: T::Assoc,
}
```

`Foo<()>` is a normalized type, but the field `x` of `Foo` has type `<() as Project>::Assoc`. To find the metadata of an ADT, we need to look at its tail and normalize these projections away. That's what the `normalize` function passed as the second parameter of `ptr_metadata_ty` is for. If the normalize function returns an unnormalized type,  then `ptr_metadata_ty` will ICE after this PR or return false for the second output before this PR.

> I think the issue might be here:
> 
> https://github.com/rust-lang/rust/blob/16ba56c24238b2f44e0a97da760c83d8795abd80/compiler/rustc_const_eval/src/interpret/terminator.rs#L686-L692
>
> We project to a field but then do not normalize again.

The types of tuple fields are always fully normalized if the tuple type itself is fully normalized. You can't have a tuple `(..., T, ...)` and the field be `<T as Project>::Assoc` or something like you can with ADTs.

Reply:
> The types of tuple fields are always fully normalized if the tuple type itself is fully normalized. 

Ah okay. We should probably have a comment there to document why we don't need to normalize.

> Even if all types are normalized, a field of a normalized ADT can still have an unnormalized type.

Okay. But that still doesn't mean that this is the right place to normalize. We should normalize wherever the unnormalized type creeps into the interpreter -- probably in the `project_field` function.

Reply:
I don't see how that reply related to what I said.

The bug is that someone is passing non-normalized types around the interpreter. Various parts of the interpreter assume things are normalized. The code in `layout_compat` is correct, some other code is wrong for hading non-normalized types to `layout_compat`.

Reply:
Oh you are saying even if we have a normalized type in `layout_compat` it can still hit non-normalized types further down. Now I see. This needs a big fat warning in the `ptr_metadata_ty` docs.

Nevertheless is still also seems wrong that `project_field` does not normalize? Or is that fine?

Reply:
We could remove the normalize parameter from `ptr_metadata_ty` entitely and always use `normalize_erasing_regions` with `ParamEnv::reveal_all()`. Or can there be any type information in the `InterpCx`'s param env that is required for this normalization?

Reply:
it's also required by typeck/borrowck where normalization must not use `normalize_erasing_regions` as we need to handle region constraints from normalization

Reply:
I moved the miri changes to https://github.com/rust-lang/rust/pull/120587.